ios/RNSScreen.mm

```diff
@@ -1087,21 +1084,56 @@ - (CGSize)getCalculatedStatusBarHeightIsModal:(BOOL)isModal
 #endif /* Check for iOS 13.0 */
 
 #else
-  // On TVOS, status bar doesn't exist
+  // TVOS does not have status bar.
   return CGSizeMake(0, 0);
 #endif // !TARGET_OS_TV
 }
 
+- (UINavigationController *)getVisibleNavigationControllerIsModal:(BOOL)isModal
+{
+  UINavigationController *navctr = self.navigationController;
+
+  if (isModal) {
+    // In case where screen is a modal, we want to calculate childViewController's
+    // navigation bar height instead of the navigation controller from RNSScreen.
+    if (self.assertModalHierarchy) {
+      navctr = self.childViewControllers[0];
+    } else {
+      // If the modal does not meet requirements (there's no RNSNavigationController which means that probably it
+      // doesn't have header or there are more than one RNSNavigationController which is invalid) we don't want to
+      // return anything.
+      return nil;
+    }
+  }
+
+  return navctr;
+}
+
 - (CGFloat)calculateHeaderHeightIsModal:(BOOL)isModal
 {
-  CGFloat navbarHeight = [self getCalculatedHeaderHeightIsModal:isModal];
-  CGSize statusBarSize = [self getCalculatedStatusBarHeightIsModal:isModal];
+  UINavigationController *navctr = [self getVisibleNavigationControllerIsModal:isModal];
+  CGSize statusBarSize = [self getStatusBarHeightIsModal:isModal];
+
+  // If navigation controller doesn't exists (or it is hidden) we want to handle two possible cases.
+  // If there's no navigation controller for the modal, we simply don't want to return header height, as modal possibly
+  // does not have header and we don't want to count status bar. If there's no navigation controller for the view we
+  // just want to return status bar height (if it's hidden, it will simply return 0).
+  if (navctr == nil || navctr.isNavigationBarHidden) {
+    if (isModal)
+      return 0;
+    else
+      return MIN(statusBarSize.width, statusBarSize.height);
+  }
```

Comment:
```suggestion
  UINavigationController *navctr = [self getVisibleNavigationControllerIsModal:isModal];

  // If navigation controller doesn't exists (or it is hidden) we want to handle two possible cases.
  // If there's no navigation controller for the modal, we simply don't want to return header height, as modal possibly
  // does not have header and we don't want to count status bar. If there's no navigation controller for the view we
  // just want to return status bar height (if it's hidden, it will simply return 0).
  if (navctr == nil || navctr.isNavigationBarHidden) {
    if (isModal)
      return 0;
    else {
      CGSize statusBarSize = [self getStatusBarHeightIsModal:isModal];
      return MIN(statusBarSize.width, statusBarSize.height);
    }
  }
```

Reply:
Resolved in https://github.com/software-mansion/react-native-screens/pull/1917/commits/add6ee3ea469736574a26cece43b2dbfcb714437.